src/bundles/analytics.js

```diff
@@ -149,19 +156,18 @@ const selectors = {
    */
   selectAnalyticsEnabled: (state) => state.analytics.consent.length > 0,
   /**
-   * Ask the user if we may enable analytics.
    * @param {State} state
    */
-  selectAnalyticsAskToEnable: (state) => {
-    const { lastEnabledAt, lastDisabledAt, consent } = state.analytics
-    // user has not explicitly chosen
-    if (!lastEnabledAt && !lastDisabledAt && consent.length === 0) {
-      // ask to enable.
-      return true
-    }
-    // user has already made an explicit choice; dont ask again.
-    return false
-  },
+  selectAnalyticsOptedOutPriorToDefaultOptIn: (state) => state.analytics.consent.length === 0 && state.analytics.lastDisabledAt > state.analytics.lastEnabledAt,
```

Comment:
After looking through the rest of the code.. I'm sure that things will work with this test, but concerned that we're not checking for either

1. `!state.analytics.optedOut`
2. or setting lastDisabledAt & lastEnabledAt to 0.

With the current code.. the only implicit blocking of this returning true again in the future is that the call to `selectAnalyticsOptedOutPriorToDefaultOptIn` is blocked by a call to `selectAnalyticsOptedOut`. However, if someone is opted out, then enables consent, then disables consent, then we will show the banner again, won't we? Is there already a test for that?

Reply:
i think we just need to have a check for optedout. `if !state.analytics.optedOut && consent === 0 && lastDisabledAt > lastEnabledAt`

| State name | lastDisabledAt | lastEnabledAt | optedOut | !optedOut && lastDisabledAt > lastEnabledAt | show banner? |
|------------|:--------------:|---------------|----------|---------------------------------------------|--------------|
| Disabled   | 2022-01-01     | 0             | false    | true                                        | true         |
| Enabled    | 2022-01-01     | 2022-01-02    | false    | false                                       | false        |
| Disabled   | 2022-01-03     | 2022-01-02    | true     | false                                       | false        |

Reply:
thank you for the detailed charts, I do believe all of these scenarios are covered because we have a check in the init that state.analytics.optedOut and within that condition if they have opted out we essentially do nothing, if they have not opted out and previously disabled (row 1 in table above) then that is the only time the showAnalyticsBanner is set to true otherwise the showAnalyticsBanner is reset to false during other actions (should maybe be moved or more explicit in the code)

I am pretty sure tests will cover all of the above scenarios but there are a lot of them! basically I used lastDisabledAt + lastEnabledAt + consent.length for all existing and previous logic, then added the two new flags optedOut and showAnalyticsBanner to cover future scenarios of users toggling on/off consent